### ISSUE TITLE
spectraData pulls mz and intenstities

### ISSUE THREAD

**Issue by lgatto** [rformassspectrometry/Spectra#92]
`spectraData` extracts by default the *mz* and *intensities*, which makes the call extremely slow and return a big object:

```r
> basename(f <- msdata::proteomics(pattern  = "20141210", full.names = TRUE))
[1] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz"
> sp <- Spectra(f, backend = MsBackendMzR())
> system.time(spd1 <- spectraData(sp, setdiff(spectraVariables(sp), c("mz", "intensity"))))
   user  system elapsed 
  0.018   0.000   0.018 
> print(object.size(spd1), units = "Mb")
2.8 Mb
> system.time(spd2 <- spectraData(sp))
   user  system elapsed 
 16.447   0.267  16.571 
> print(object.size(spd2), units = "Mb")
576.1 Mb
```

I would suggest, at least for on-disk backends, that mz and intensity should not be part of the default spectraVariables. In these cases, we could add

```r
spectraData(sp, mz = FALSE, intensity = FALSE)
```

as defaults, that get set to TRUE when these are provided explicitly as part of the `spectraVariables` argument.

Note however that this is dramatically reduced when using the hfd5 backend.

```r
> sp2 <- setBackend(sp, backend = MsBackendHdf5Peaks())
> system.time(spd3 <- spectraData(sp2, setdiff(spectraVariables(sp), c("mz", "intensity"))))
   user  system elapsed 
  0.015   0.000   0.015 
> print(object.size(spd3), units = "Mb")
2.8 Mb
> system.time(spd4 <- spectraData(sp2))
   user  system elapsed 
  4.673   0.188   4.973 
> print(object.size(spd4), units = "Mb")
576.1 Mb
```

@jorainer @sgibb Comments, suggestions?

Cc @plantton

**Comment by jorainer** [comment]
Honestly, I don't like the idea to add `mz` and `intensity` as dedicated parameters to the function. If users want specific columns they can use the `columns` parameter. Also, `spectraData` was not intended to be called often - it's more a convenience function to extract all the data as a `DataFrame`. In day-to-day use users should use the `Spectra` object itself - it can be used like a `DataFrame` anyway, i.e. it is possible to access any spectra variable with `$`.

**Comment by lgatto** [comment]
I hear you regarding additional arguments to `spectraData`. But I guess my question is whether we really want the *mz* and *intensity* columns to be pulled out with the spectra (meta-data)? This isn't making any sense for on disk backends.

(As for day-to-day use, I will follow up in another issue that lead to this one, and the extraction of the `spectraData` return value.)

**Comment by sgibb** [comment]
Initially the `spectraData` was meant to be an accessor for the metadata (`header` in `mzR`). IMHO the `MsBackendDataFrame` misuses the metadata to store its spectra data. While that's ok from the developer perspective the user should not use `spectraData(x)` to access mz/intensity values. We should separate metadata and spectra data in the user visible API. 
So I agree with @lgatto that `spectraData` should not return `mz` and `intensity` values.

If the user really wants to have all metadata and mz/intensity values as `DataFrame` we should provide an `as(x, "DataFrame")`/`as.data.frame` method (would be maybe useful for plotting with `ggplot` anyway).

**Comment by jorainer** [comment]
I see your points. Currently, the main purpose of the `spectraData` (in the `MsBackend`!)) is to have a way to move data from one to another backend. The end user should ideally not use this method, so maybe it's better we drop the `spectraData,Spectra` method alltogether? We could have a `spectraMetaData` method for `Spectra` that returns just the data for all spectra variables except m/z and intensity values? Could that be a compromise @lgatto and @sgibb ?

If possible I would avoid to change the `spectraData` and `spectraData<-` methods for `MsBackend` as that would involve quite some code rewriting.

**Comment by lgatto** [comment]
I think the `spectraData` is a very important function. It is the equivalent of `colData`/`fData`, and we definitely need to keep it. I also think we had a discussion about naming, and `spectraMetaData` was possibly one of the possibilities. But I definitely prefer what we have.

One possibility would be for `spectraVariables`, that 

> returns a ‘character’ vector with the available spectra variables (columns, fields or attributes) available in ‘object’.

no to include `mz` and `intensities` (which would be more accurate IMHO anyway). Then, `spectraData`, that uses `spectraVariables` to define the columns to return, would work *as expected*. If needed, one could still get everything with

```r
spectraData(object, columns = c(spectraVariables(object), "mz", "intensity"))
```

**Comment by jorainer** [comment]
Sure, that's the simplest solution. I'm OK with it.

**Comment by lgatto** [comment]
PR #94  submitted

**Comment by lgatto** [comment]
There are quite a few places in the unit tests that reply on `spectraData()` to return the `mz` and `intensity` columns. I am going to add an `as(, "DataFrame")` coercion method as a replacement for this.

**Comment by lgatto** [comment]
Currently being a bit crushed under the effects of removing these `mz` and `intensity` from the default  `spectraVariables`, it looks like `spectraData` was used quite a bit throughout the code base to extract `mz` and `intensities` with many ramifications. This proves a bit disruptive than expected.

**Comment by jorainer** [comment]
Yes, the `spectraData,MsBackend` is one of the main core functions used e.g. when changing backends, to *lift-over* all data from one to the other backend. But the methods for `MsBackend` should not be directly called by the user. Shouldn't it suffice to change the default for `Spectra`? Or better said, change the `spectraVariables,Spectra` method to remove `"mz"` and `"intensity"` from the result called on `spectraVariables(object@backend)`?

**Comment by lgatto** [comment]
No, I thought it would be that easy, but lots of tests suddenly fail.

**Comment by jorainer** [comment]
Suggestion:

```r
setMethod("spectraVariables", "Spectra", function(object) {
    svars <- spectraVariables(object@backend)
    svars[!(svars %in% c("mz", "intensity"))]
})

```

**Comment by lgatto** [comment]
Current plans:
- Have `spectraData,Spectra` to not return `mz` and  `intensity`.
- Rename all `spectraData` backend methods to `asDataFrame`, which would return `mz` and  `intensity`.

**Comment by jorainer** [comment]
I started giving it a go - some first things: we will have two new methods (might make sense to add them to `ProtGenerics` because we will need them also for `Chromatograms`):
- `asDataFrame,MsBackend` to get all data.
- `setDataFrame<-,MsBackend` to set all data.

Also, we have to change the parameter `spectraData` in the `backendInitialize,MsBackendDataFrame` - I would suggest to change that to `data`.

I am currently hacking in the *jomaster* branch.

**Comment by lgatto** [comment]
- Did you really mean `setDataFrame<-,MsBackend`, or `asDataFrame<-,MsBackend`?

- Ok to rename the argument in `backendInitialize,MsBackendDataFrame` to `data` - I think this will actually be clearer.

- Could you push *jomaster*, please - I don't think it's available here.

**Comment by jorainer** [comment]
Before pushing I wanted to have it somewhat at a stage that does not completely fail.
And yes, I meant to use `setDataFrame<-,MsBackend` instead of `asDataFrame<-,MsBackend` as a replacement for `spectraData<-,MsBackend`. The `asDataFrame<-` does not seem to be intuitive to me.

**Comment by jorainer** [comment]
OK, I've pushed to *jomaster* (https://github.com/rformassspectrometry/Spectra/tree/jomaster)
the code *should* be working, the errors I get are, I guess, because my `devtools::document()` has some problems parsing the roxygen documentation (`@section Use of block quotes is not currently supported` - which I absolutely don't understand).

I might give it another try today evening - or in case you can give it a shot @lgatto

**Comment by jorainer** [comment]
Regarding the roxygen problem - I found that. It's because there are some lines starting with a `>`. I removed that and check the results of the unit tests again...

**Comment by jorainer** [comment]
I'm still working on it @lgatto - got a strange error, so I'm reverting to `asDataFrame,MsBacken<-` to see if I can get rid of it.

**Comment by lgatto** [comment]
Fixed by #95
